File: Data Structures & Algorithms/implement-prefix-tree/submission-9.java
class PrefixTree {
    private static final int K = 26;

    private final Node root;

    public PrefixTree() {
        this.root = new Node();
    }

    public void insert(String word) {
        Node cur = this.root;
        for (char c: word.toCharArray()) {
            int index = c - 'a';
            if (cur.children[index] == null) {
                cur.children[index] = new Node();
            }
            cur = cur.children[index];
        }
        cur.isTerminal = true;
    }

    public boolean search(String word) {
        Node cur = this.root;
        for (char c: word.toCharArray()) {
            int index = c - 'a';
            if (cur.children[index] == null) {
                return false;
            }
            cur = cur.children[index];
        }
        return cur.isTerminal;
    }

    public boolean startsWith(String prefix) {
        Node cur = this.root;
        for (char c: prefix.toCharArray()) {
            int index = c - 'a';
            if (cur.children[index] == null) {
                return false;
            }
            cur = cur.children[index];
        }
        return true;
    }

    private static class Node {
        private final Node[] children;

        private boolean isTerminal;

        public Node() {
            this.children = new Node[26];
            this.isTerminal = false;
        }
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-9.rs
struct TrieNode {
    children: [Option<Box<TrieNode>>; 26],
    word: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: Default::default(),
            word: false,
        }
    }
}

struct WordDictionary {
    root: TrieNode,
}

impl WordDictionary {
    fn new() -> Self {
        WordDictionary { root: TrieNode::new() }
    }

    fn add_word(&mut self, word: String) {
        let mut cur = &mut self.root;
        for c in word.bytes() {
            let idx = (c - b'a') as usize;
            cur = cur.children[idx].get_or_insert_with(|| Box::new(TrieNode::new()));
        }
        cur.word = true;
    }

    fn search(&self, word: String) -> bool {
        Self::dfs(word.as_bytes(), 0, &self.root)
    }

    fn dfs(word: &[u8], j: usize, root: &TrieNode) -> bool {
        let mut cur = root;
        for i in j..word.len() {
            let c = word[i];
            if c == b'.' {
                for child in &cur.children {
                    if let Some(node) = child {
                        if Self::dfs(word, i + 1, node) {
                            return true;
                        }
                    }
                }
                return false;
            } else {
                let idx = (c - b'a') as usize;
                match &cur.children[idx] {
                    None => return false,
                    Some(node) => cur = node,
                }
            }
        }
        cur.word
    }
}